go/tendermint/apps/staking/state/state.go

```diff
@@ -543,6 +565,64 @@ func (s *MutableState) TransferFromCommon(ctx *abci.Context, toID signature.Publ
 	return ret, nil
 }
 
+// AddRewards computes and transfers the staking rewards to active escrow accounts.
+// If an error occurs, the pool and affected accounts are left in an invalid state.
+func (s *MutableState) AddRewards(time epochtime.EpochTime, factor *staking.Quantity, accounts map[signature.MapKey]bool) error {
+	steps, err := s.RewardSchedule()
+	if err != nil {
+		return err
+	}
+	var activeStep *staking.RewardStep
+	for _, step := range steps {
+		if time < step.Until {
+			activeStep = &step
+			break
+		}
+	}
+	if activeStep == nil {
+		// We're past the end of the schedule.
+		return nil
+	}
+
+	commonPool, err := s.CommonPool()
+	if err != nil {
+		return errors.Wrap(err, "loading common pool")
+	}
+
+	for mk := range accounts {
+		var id signature.PublicKey
+		id.FromMapKey(mk)
+
+		ent := s.Account(id)
+
+		q := ent.Escrow.Active.Balance.Clone()
+		// Multiply first.
+		if err := q.Mul(factor); err != nil {
+			return errors.Wrap(err, "multiplying by reward factor")
+		}
+		if err := q.Mul(&activeStep.Scale); err != nil {
+			return errors.Wrap(err, "multiplying by reward step scale")
+		}
+		if err := q.Quo(staking.RewardAmountDenominator); err != nil {
+			return errors.Wrap(err, "dividing by reward amount denominator")
+		}
+
+		if q.IsZero() {
+			continue
+		}
+
+		if err := staking.Move(&ent.Escrow.Active.Balance, commonPool, q); err != nil {
+			return errors.Wrap(err, "transferring to active escrow balance from common pool")
+		}
+
+		s.SetAccount(id, ent)
```

Comment:
We should do all updates at the end when we are sure that there are no errors (or use a snapshot?). Otherwise this can cause stuff to still be committed (e.g., there is no implicit rollback in case of errors).

Note that in the current implementation this is probably not a problem as any error here will cause a consensus failure and nothing will be committed. But in case this is ever used outside of `BeginBlock` it may be useful to mitigate.

Another reason for deferring updates is non-determinism -- IAVL is sensitive to order of updates, so we (currently) need to sort the updates before committing them. _(An alternative is to have an intermediate layer that caches and sorts all IAVL updates before they are committed or to use our own data structure, e.g. #1898, which doesn't have this problem. But this is out of scope for this PR.)_

Reply:
is there a place outside of BeginBlock that would disrespect an error condition and proceed to act on the same tree? I'd consider continuing in that condition to be the problem.

I'll treat the nondeterministic map iteration as a separate issue and discuss that in your other comment.

Reply:
> is there a place outside of BeginBlock that would disrespect an error condition and proceed to act on the same tree? I'd consider continuing in that condition to be the problem.

Potentially in DeliverTx. An error there aborts the transaction but should not commit any state changes.

We could enforce that at the mux layer, but then we need a way to still have some state side effects even for failing transactions (because maybe sometimes you would want some even then).

Reply:
> aborts the transaction but should not commit any state changes

aw ew

probably better to snapshot, so that we have slightly less "have all accounts in memory" situations, in case we ever have to get rid of them

Reply:
Yeah makes sense. So for now I see that you just have a comment that says that accounts may be corrupted in case of errors. Maybe just add something like "if this is an issue, the caller should snapshot and rollback in case of errors".

Reply:
added